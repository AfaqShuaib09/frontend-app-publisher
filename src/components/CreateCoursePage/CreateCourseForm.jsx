import React from 'react';
import { Field, reduxForm } from 'redux-form';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { push } from 'connected-react-router';

import RenderInputTextField from '../RenderInputTextField';
import RenderSelectField from '../RenderSelectField';
import store from '../../data/store';
import { AUDIT_TRACK, VERIFIED_TRACK, PROFESSIONAL_TRACK } from '../../data/constants';

class CreateCourseForm extends React.Component {
  getEnrollmentTrackOptions() {
    return [
      { label: 'Select enrollment track', value: '' },
      { label: VERIFIED_TRACK.name, value: VERIFIED_TRACK.key },
      { label: AUDIT_TRACK.name, value: AUDIT_TRACK.key },
      { label: PROFESSIONAL_TRACK.name, value: PROFESSIONAL_TRACK.key },
    ];
  }

  processOrganizations(organizations) {
    const orgSelectList = [{ label: 'Select user organization', value: '' }];

    if (organizations) {
      organizations.forEach((org) => {
        orgSelectList.push({ label: org.name, value: org.key });
      });
    }
    return orgSelectList;
  }

  render() {
    const {
      handleSubmit,
      organizations,
      pristine,
      submitting,
    } = this.props;

    return (
      <React.Fragment>
        <div className="create-course-form row">
          <div className="col">
            <form onSubmit={handleSubmit}>
              <Field
                name="courseOrg"
                component={RenderSelectField}
                options={this.processOrganizations(organizations)}
                label={
                  <React.Fragment>
                    Organization
                    <span className="required">*</span>
                  </React.Fragment>
                }
                required
              />
              <Field
                name="courseTitle"
                component={RenderInputTextField}
                type="text"
                label={
                  <React.Fragment>
                    Course title
                    <span className="required">*</span>
                  </React.Fragment>
                }
                required
              />
              <Field
                name="courseNumber"
                component={RenderInputTextField}
                type="text"
                label={
                  <React.Fragment>
                    Course number
                    <span className="required">*</span>
                  </React.Fragment>
                }
                required
              />
              <Field
                name="courseEnrollmentTrack"
                component={RenderSelectField}
                options={this.getEnrollmentTrackOptions()}
                label={
                  <React.Fragment>
                    Enrollment track
                    <span className="required">*</span>
                  </React.Fragment>
                }
                required
              />
              <Field
                name="coursePrice"
                component={RenderInputTextField}
                type="number"
                label={
                  <React.Fragment>
                    Course price (USD)
                    <span className="required">*</span>
                  </React.Fragment>
                }
                required
              />
              <div className="row justify-content-end">
                <Link
                  className={['btn btn-link ml-3 form-cancel-btn']}
                  to="/"
                  disabled={pristine || submitting}
                  onClick={() => {
                    store.dispatch(push('/'));
                  }}
                >
                  Cancel
                </Link>
                <button type="submit" className="form-submit-btn" disabled={pristine || submitting}>
                  Submit
                </button>
              </div>
            </form>
          </div>
        </div>
      </React.Fragment>
    );
  }
}

CreateCourseForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  initialValues: PropTypes.shape({ // eslint-disable-line react/no-unused-prop-types
    courseOrg: PropTypes.string,
    courseTitle: PropTypes.string,
    courseNumber: PropTypes.string,
    courseEnrollmentTrack: PropTypes.string,
    coursePrice: PropTypes.number,
  }).isRequired,
  organizations: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    key: PropTypes.string.isRequired,
  })).isRequired,
  pristine: PropTypes.bool.isRequired,
  submitting: PropTypes.bool.isRequired,
};

export default reduxForm({ form: 'create-course-form' })(CreateCourseForm);

