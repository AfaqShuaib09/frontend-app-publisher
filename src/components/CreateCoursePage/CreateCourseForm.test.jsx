import React from 'react';
import { shallow } from 'enzyme';

import { BaseCreateCourseForm } from './CreateCourseForm';

const organizations = [
  { name: 'edX', key: 'edx' }, { name: 'edX2', key: 'edx2' },
];

describe('CreateCourseForm', () => {
  const initialValues = {
    org: 'edx',
    title: 'Hello',
    number: 'edx101',
    enrollmentTrack: 'verified',
    price: 100.00,
  };

  it('renders html correctly with no orgs', () => {
    const component = shallow(<BaseCreateCourseForm
      handleSubmit={() => {}}
      initialValues={{}}
      organizations={organizations}
    />);
    expect(component).toMatchSnapshot();
  });

  it('renders html correctly with data', () => {
    const component = shallow(<BaseCreateCourseForm
      handleSubmit={() => {}}
      initialValues={initialValues}
      currentFormValues={initialValues}
      organizations={organizations}
    />);
    expect(component).toMatchSnapshot();
  });

  it('renders html correctly while submitting', () => {
    const component = shallow(<BaseCreateCourseForm
      submitting
      isCreating
      pristine={false}
      handleSubmit={() => {}}
      initialValues={{}}
      currentFormValues={initialValues}
      organizations={organizations}
    />);
    expect(component).toMatchSnapshot();
  });
});
