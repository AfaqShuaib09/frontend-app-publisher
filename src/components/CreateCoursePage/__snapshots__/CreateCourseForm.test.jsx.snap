// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateCourseForm renders html correctly while submitting 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ReduxForm
    handleSubmit={[Function]}
    initialValues={
      Object {
        "courseEnrollmentTrack": "",
        "courseNumber": "",
        "courseOrg": "",
        "coursePrice": "",
        "courseTitle": "",
      }
    }
    organizations={
      Array [
        Object {
          "key": "edx",
          "name": "edX",
        },
        Object {
          "key": "edx2",
          "name": "edX2",
        },
      ]
    }
    submitting={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "destroyOnUnmount": true,
      "enableReinitialize": false,
      "forceUnregisterOnUnmount": false,
      "form": "create-course-form",
      "getFormState": [Function],
      "handleSubmit": [Function],
      "initialValues": Object {
        "courseEnrollmentTrack": "",
        "courseNumber": "",
        "courseOrg": "",
        "coursePrice": "",
        "courseTitle": "",
      },
      "keepDirtyOnReinitialize": false,
      "organizations": Array [
        Object {
          "key": "edx",
          "name": "edX",
        },
        Object {
          "key": "edx2",
          "name": "edX2",
        },
      ],
      "persistentSubmitErrors": false,
      "pure": true,
      "shouldAsyncValidate": [Function],
      "shouldError": [Function],
      "shouldValidate": [Function],
      "shouldWarn": [Function],
      "submitting": true,
      "touchOnBlur": true,
      "touchOnChange": false,
      "updateUnregisteredFields": false,
    },
    "ref": [Function],
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "destroyOnUnmount": true,
        "enableReinitialize": false,
        "forceUnregisterOnUnmount": false,
        "form": "create-course-form",
        "getFormState": [Function],
        "handleSubmit": [Function],
        "initialValues": Object {
          "courseEnrollmentTrack": "",
          "courseNumber": "",
          "courseOrg": "",
          "coursePrice": "",
          "courseTitle": "",
        },
        "keepDirtyOnReinitialize": false,
        "organizations": Array [
          Object {
            "key": "edx",
            "name": "edX",
          },
          Object {
            "key": "edx2",
            "name": "edX2",
          },
        ],
        "persistentSubmitErrors": false,
        "pure": true,
        "shouldAsyncValidate": [Function],
        "shouldError": [Function],
        "shouldValidate": [Function],
        "shouldWarn": [Function],
        "submitting": true,
        "touchOnBlur": true,
        "touchOnChange": false,
        "updateUnregisteredFields": false,
      },
      "ref": [Function],
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`CreateCourseForm renders html correctly with data 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ReduxForm
    handleSubmit={[Function]}
    initialValues={
      Object {
        "courseEnrollmentTrack": "verified",
        "courseNumber": "edx101",
        "courseOrg": "edx",
        "coursePrice": 100,
        "courseTitle": "Hello",
      }
    }
    organizations={
      Array [
        Object {
          "key": "edx",
          "name": "edX",
        },
        Object {
          "key": "edx2",
          "name": "edX2",
        },
      ]
    }
    pristine={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "destroyOnUnmount": true,
      "enableReinitialize": false,
      "forceUnregisterOnUnmount": false,
      "form": "create-course-form",
      "getFormState": [Function],
      "handleSubmit": [Function],
      "initialValues": Object {
        "courseEnrollmentTrack": "verified",
        "courseNumber": "edx101",
        "courseOrg": "edx",
        "coursePrice": 100,
        "courseTitle": "Hello",
      },
      "keepDirtyOnReinitialize": false,
      "organizations": Array [
        Object {
          "key": "edx",
          "name": "edX",
        },
        Object {
          "key": "edx2",
          "name": "edX2",
        },
      ],
      "persistentSubmitErrors": false,
      "pristine": true,
      "pure": true,
      "shouldAsyncValidate": [Function],
      "shouldError": [Function],
      "shouldValidate": [Function],
      "shouldWarn": [Function],
      "touchOnBlur": true,
      "touchOnChange": false,
      "updateUnregisteredFields": false,
    },
    "ref": [Function],
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "destroyOnUnmount": true,
        "enableReinitialize": false,
        "forceUnregisterOnUnmount": false,
        "form": "create-course-form",
        "getFormState": [Function],
        "handleSubmit": [Function],
        "initialValues": Object {
          "courseEnrollmentTrack": "verified",
          "courseNumber": "edx101",
          "courseOrg": "edx",
          "coursePrice": 100,
          "courseTitle": "Hello",
        },
        "keepDirtyOnReinitialize": false,
        "organizations": Array [
          Object {
            "key": "edx",
            "name": "edX",
          },
          Object {
            "key": "edx2",
            "name": "edX2",
          },
        ],
        "persistentSubmitErrors": false,
        "pristine": true,
        "pure": true,
        "shouldAsyncValidate": [Function],
        "shouldError": [Function],
        "shouldValidate": [Function],
        "shouldWarn": [Function],
        "touchOnBlur": true,
        "touchOnChange": false,
        "updateUnregisteredFields": false,
      },
      "ref": [Function],
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`CreateCourseForm renders html correctly with no orgs 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ReduxForm />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "destroyOnUnmount": true,
      "enableReinitialize": false,
      "forceUnregisterOnUnmount": false,
      "form": "create-course-form",
      "getFormState": [Function],
      "initialValues": undefined,
      "keepDirtyOnReinitialize": false,
      "persistentSubmitErrors": false,
      "pure": true,
      "shouldAsyncValidate": [Function],
      "shouldError": [Function],
      "shouldValidate": [Function],
      "shouldWarn": [Function],
      "touchOnBlur": true,
      "touchOnChange": false,
      "updateUnregisteredFields": false,
    },
    "ref": [Function],
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "destroyOnUnmount": true,
        "enableReinitialize": false,
        "forceUnregisterOnUnmount": false,
        "form": "create-course-form",
        "getFormState": [Function],
        "initialValues": undefined,
        "keepDirtyOnReinitialize": false,
        "persistentSubmitErrors": false,
        "pure": true,
        "shouldAsyncValidate": [Function],
        "shouldError": [Function],
        "shouldValidate": [Function],
        "shouldWarn": [Function],
        "touchOnBlur": true,
        "touchOnChange": false,
        "updateUnregisteredFields": false,
      },
      "ref": [Function],
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
